### PR TITLE
Added `PlutusTypeDataList` safer deriving strat for `ProductisData`

### PR DESCRIPTION
**Pull request by SeungheonOh**
`PlutusTypeDataList` enforces given types' internal to be
`PDataRecord`.

Plutonomicon/plutarch-plutus#564 will make it derive `PDataField`.

### CODE REVIEW

**Review by kozross — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
...
 This format is based on [Keep A Changelog](https://keepachangelog.com/en/1.0.0).
 
+## 3.4.0 -- 2022-08-30
+
+### Added
+
+* `PlutusTypeDataList` is a new deriving strategy for `ProductIsData`. It will
+  only accept Plutarch types that have `PDataRecord` inner. Otherwise, it behaves
+  identically to `PlutusTypeNewtype`.
+
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply CHANGELOG.md:12]: got it.

**Comment by SeungheonOh** [comment]
```hs
data Foo
    = Foo
    { hello :: Integer
    , world :: Integer
    }
    deriving stock (Generic)
    deriving (ToData, FromData) via (ProductIsData Foo)
    deriving (PConstantDecl) via (DerivePConstantViaDataList Foo PFoo)
    deriving anyclass (SOP.Generic)

instance PUnsafeLiftDecl PFoo where type PLifted PFoo = Foo

newtype PFoo s
    = PFoo
      ( Term s
          ( PDataRecord
              '[ "abc" ':= PInteger
               , "asd" ':= PInteger
               , "efg" ':= PInteger               
               ]
          )
      )
```

I made it so that it will throw up when you give unmatching types or number of fields. So following will error like this:

```
src/Plutarch/Extra/IsData.hs:99:10-30: error:
    • More/Less types are given to the Plutarch type
      '[PInteger, PInteger]
      '[PInteger, PInteger, PInteger]
    • In the instance declaration for ‘DerivePlutusType PFoo’
   |
99 | instance DerivePlutusType PFoo where
```

**Review by chfanghr — APPROVED** [review]

**Comment by peter-mlabs** [comment]
Can we get a usage example in the haddock?

**Comment by SeungheonOh** [comment]
Good idea

File: src/Plutarch/Extra/IsData.hs [hunk 2]
```diff
@@ -108,6 +106,65 @@ newtype ProductIsData (a :: Type) = ProductIsData {unProductIsData :: a}
...
+    MatchTypes' '[] ys = 'False
+    MatchTypes' xs '[] = 'False
+
+type family MatchTypesError (n :: [[Type]]) (m :: [S -> Type]) (a :: Bool) :: Constraint where
+    MatchTypesError _ _ 'True = ()
+    MatchTypesError n m 'False =
+        ( 'True ~ 'False
+        , TypeError
+            ( 'Text "More/Less types are given to the Plutarch type"
+                ':$$: 'ShowType (GetRecordTypes n)
```
Comment (peter-mlabs) [inline src/Plutarch/Extra/IsData.hs:136]: Could we get `Constituent Haskell Types:` and `Constituent Plutarch Types:` annotation here? It won't be apparent unless you've familiar with the types.

Also, can we make the `"More/Less types are given to the Plutarch type"` more descriptive? Maybe something like `"Error in 'PlutusTypeDataList': Mismatch between constituent Haskell and Plutarch types"`

Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:136]: It'd be nice. Feel free to commit it yourself, because I'll have access to my desktop in few hours.

Comment (peter-mlabs) [reply src/Plutarch/Extra/IsData.hs:136]: let me figure out how XD

Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:136]: it's just like string : D

Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs]: I did it.

File: src/Plutarch/Extra/IsData.hs [hunk 3]
```diff
@@ -72,34 +61,46 @@ import PlutusTx (
...
-{- |
-  Wrapper for deriving 'ToData', 'FromData' using the 'List' constructor of Data to represent a Product type.
-
-  Uses 'gProductToBuiltinData', 'gproductFromBuiltinData'.
+{- | Wrapper for deriving 'ToData', 'FromData' using the 'List'
+     constructor of Data to represent a Product type.
+
+     It is recommended to use `PlutusTypeDataList` when deriving
+     `PlutusType` as it provides some basic safety by ensuring
+     Plutarch types have an `Inner` type of `PDataRecord`.
```
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:73]: These should use `'`-wrapping so the Haddocks link. If they're not linkable, `@`-wrapping is what you want for monospacing. Unfortunately, Haddock syntax is a bit wonky.

File: src/Plutarch/Extra/IsData.hs [hunk 4]
```diff
@@ -72,34 +61,46 @@ import PlutusTx (
...
+
+     Uses 'gProductToBuiltinData', 'gproductFromBuiltinData'.
+
+ = Example
+@
+data Foo =
+  Foo Integer [Integer]
+  deriving stock (Generic)
+  deriving anyclass (SOP.Generic)
```
Comment (kozross) [inline src/Plutarch/Extra/IsData.hs:82]: Add a qualified import to the top of the example to tell people what `SOP` refers to.

**Review by kozross — CHANGES_REQUESTED** [review]

File: src/Plutarch/Extra/IsData.hs (hunk 3 above)
Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:73]: what does ``` do?

Comment (kozross) [reply src/Plutarch/Extra/IsData.hs:73]: I actually don't think it does anything?

Comment (SeungheonOh) [reply src/Plutarch/Extra/IsData.hs:73]: :O

**Review by kozross — APPROVED** [review]

**Review by kozross — CHANGES_REQUESTED** [review]
Cabal file version needs to be raised.

**Review by kozross — APPROVED** [review]

**Pull request merged**